Framework/Algorithms/src/ApplyTransmissionCorrection.cpp

```diff
@@ -81,61 +118,14 @@ void ApplyTransmissionCorrection::exec() {
     ETrIn = transWS->readE(0);
   }
 
-  const int numHists = static_cast<int>(inputWS->getNumberHistograms());
-  Progress progress(this, 0.0, 1.0, numHists);
-
   // Create a Workspace2D to match the intput workspace
   MatrixWorkspace_sptr corrWS = WorkspaceFactory::Instance().create(inputWS);
 
-  // Loop through the spectra and apply correction
-  PARALLEL_FOR2(inputWS, corrWS)
-  for (int i = 0; i < numHists; i++) {
-    PARALLEL_START_INTERUPT_REGION
-
-    IDetector_const_sptr det;
-    try {
-      det = inputWS->getDetector(i);
-    } catch (Exception::NotFoundError &) {
-      g_log.warning() << "Spectrum index " << i
-                      << " has no detector assigned to it - discarding"
-                      << std::endl;
-      // Catch if no detector. Next line tests whether this happened - test
-      // placed
-      // outside here because Mac Intel compiler doesn't like 'continue' in a
-      // catch
-      // in an openmp block.
-    }
-    // If no detector found, skip onto the next spectrum
-    if (!det)
-      continue;
-
-    // Copy over the X data
-    corrWS->dataX(i) = inputWS->readX(i);
-
-    // Skip if we have a monitor or if the detector is masked.
-    if (det->isMonitor() || det->isMasked())
-      continue;
-
-    // Compute theta-dependent transmission term for each wavelength bin
-    MantidVec &YOut = corrWS->dataY(i);
-    MantidVec &EOut = corrWS->dataE(i);
-
-    const double exp_term =
-        (1.0 / cos(inputWS->detectorTwoTheta(det)) + 1.0) / 2.0;
-    for (int j = 0; j < static_cast<int>(inputWS->readY(0).size()); j++) {
-      if (!thetaDependent) {
-        YOut[j] = 1.0 / TrIn[j];
-        EOut[j] = std::fabs(ETrIn[j] * TrIn[j] * TrIn[j]);
-      } else {
-        EOut[j] = std::fabs(ETrIn[j] * exp_term / pow(TrIn[j], exp_term + 1.0));
-        YOut[j] = 1.0 / pow(TrIn[j], exp_term);
-      }
-    }
-
-    progress.report("Applying Transmission Correction");
-    PARALLEL_END_INTERUPT_REGION
-  }
-  PARALLEL_CHECK_INTERUPT_REGION
+  ApplyTransmissionCorrectionToSpectrum apply(TrIn, ETrIn, thetaDependent);
+  this->for_each(
```

Comment:
You are right about the naming, there are also other things I do not like (`SpectrumAlgorithm`?). To avoid clashes, maybe we can put them into a class? Or just forbid using `using namespace Getters;`?

I like the combination idea, we just have to figure out a nice way of combining them with potential further getters into a single tuple (or just go with nested tuples?).